scheduler/src/cook/config_incremental.clj

```diff
@@ -0,0 +1,109 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns cook.config-incremental
+  (:require [clojure.tools.logging :as log]
+            [cook.datomic :as datomic]
+            [datomic.api :as d]
+            [plumbing.core :refer [for-map map-from-vals]])
+  (:import (java.util UUID)))
+
+(defn get-conn
+  "Get datomic database connection"
+  []
+  datomic/conn)
+
+(defn incremental-values->incremental-value-ents
+  "Convert plain map incremental values to datomic entities"
+  [incremental-values]
+  (map-indexed
+    (fn [index {:keys [value portion]}]
+      {:incremental-value/ordinal index
+       :incremental-value/value value
+       :incremental-value/portion portion})
+    incremental-values))
+
+(defn write-config
+  "Write an incremental configuration to the database."
+  [key incremental-values]
+  @(d/transact
+     (get-conn)
+     [; We need to retract existing incremental values; otherwise,
+      ; values simply get added to the existing list
+      [:db.fn/resetAttribute
+       [:incremental-configuration/key key]
+       :incremental-configuration/incremental-values
+       nil]
+      {:db/id (d/tempid :db.part/user)
+       :incremental-configuration/key key
+       :incremental-configuration/incremental-values (incremental-values->incremental-value-ents incremental-values)}]))
+
+(defn incremental-value-ents->incremental-values
+  "Convert datomic incremental value entities to plain maps"
+  [incremental-value-ents]
+  (map (fn [{:keys [:incremental-value/value :incremental-value/portion]}]
+         {:value value :portion portion})
+       incremental-value-ents))
```

Comment:
Don't we need to sort by :incremental-value/ordinal somewhere in here?

Reply:
Maybe it would be more straightforward to have the sort here instead of in both `read-config` and `read-all-configs`?

Reply:
yeah, will update